zkevm-circuits/src/evm_circuit/execution/balance.rs

```diff
@@ -1,36 +1,37 @@
 use crate::{
     evm_circuit::{
         execution::ExecutionGadget,
+        param::N_BYTES_ACCOUNT_ADDRESS,
         step::ExecutionState,
         util::{
             common_gadget::SameContextGadget,
             constraint_builder::{
                 ConstrainBuilderCommon, EVMConstraintBuilder, ReversionInfo, StepStateTransition,
                 Transition::Delta,
             },
-            math_gadget::IsZeroGadget,
-            not, select, CachedRegion, Cell, Word,
+            math_gadget::IsZeroWordGadget,
+            not, select, AccountAddress, CachedRegion, Cell,
         },
         witness::{Block, Call, ExecStep, Transaction},
     },
     table::{AccountFieldTag, CallContextFieldTag},
     util::{
-        word::{Word32Cell, WordExpr},
+        word::{Word, Word32Cell, WordCell, WordExpr},
         Expr,
     },
 };
-use eth_types::{evm_types::GasCost, Field, ToLittleEndian};
+use eth_types::{evm_types::GasCost, Field, ToLittleEndian, ToWord};
 use halo2_proofs::{circuit::Value, plonk::Error};
 
 #[derive(Clone, Debug)]
 pub(crate) struct BalanceGadget<F> {
     same_context: SameContextGadget<F>,
-    address: Word<F>,
+    address: AccountAddress<F>,
```

Comment:
Pretty agree with this optimisation to have clear definition, and `ByteDecomposition` is a good naming with hardcode 256 base. 

I am thinking a design problem. For `wordXX`, (XX means number of limbs) cell, the calling of assign_<type> make sense, because we have loosely requirements: `number of bytes` % `number of limbs`. For `ByteDecomposition`, `number of bytes` need to strictly equals with `number of limbs`. Therefore, it will be nice to have a restriction usage by type system itself. What I means is

```
- ByteDecomposition<F, 8> only have function assign(value: u64)
- ByteDecomposition <F, 20> only have function assign(value: H160)
- ... (so as other type)
```

strictly one-to-one mapping. It's to avoid some mal-usage like `ByteDecomposition<F, 8>.assign_h160(value: H160)` because this pair really not making any sense, so intuitively maybe here shouldn't apply generic, instead it should leverage `macro` system. 
@ed255 do you think it's an over-design or is there any other simple way to leverage type system to make it works? :)

Reply:
> Making sure we don't call the assignment function with a value bigger than the type can hold sounds great! Maybe we can skip the `macros` and just add some asserts like:
> 
> ```rust
> fn assign_h160(...) {
>     assert!(N >= 20);
> }
> ```
> 
> We would get a runtime panic when the assign is used incorrectly.

(Trying the rephrase in case not get my points)

I think in this case we need to be exactly equal, right? `assert!(N == 20)` and if it's true, actually assign_h160 is a bit verbose in `ByteDecomposition` instead of just `assign` because

1. assign(word: H160) imply ByteDecomposition <F, 20>
2. assign(word: U64) imply ByteDecomposition <F, 8>

Thats why I think it will be nice to have compiler time check and catch error like `ByteDecomposition <F, 20>.assign(u64)` But seems there is no better way to do that

Reply:
> > Making sure we don't call the assignment function with a value bigger than the type can hold sounds great! Maybe we can skip the `macros` and just add some asserts like:
> > ```rust
> > fn assign_h160(...) {
> >     assert!(N >= 20);
> > }
> > ```
> > We would get a runtime panic when the assign is used incorrectly.
> 
> (Trying the rephrase in case not get my points)
> 
> I think in this case we need to be exactly equal, right? `assert!(N == 20)` and if it's true, actually assign_h160 is a bit verbose in `ByteDecomposition` instead of just `assign` because
> 
>     1. assign(word: H160) imply ByteDecomposition <F, 20>
>     2. assign(word: U64) imply ByteDecomposition <F, 8>
> 
> 
> Thats why I think it will be nice to have compiler time check and catch error like `ByteDecomposition <F, 20>.assign(u64)` But seems there is no better way to do that

Right, I think checking equality sounds better!

I agree that it looks redundant, because the N tells you already which is the compatible type, but you still need to write it explicitly.  Nevertheless I think there's no nice solution in Rust to implement the same function with different types.  I know you could do traits like `AssignU64`, `AssignH160` with `assign(word: H160)` and `assign(word: U64)` methods; but then we need to import the traits (which is a bit annoying) and if we import multiple types we need to do things like:
`<value as AssignU64>.assign(u64)` which are a bit ugly.

So implementing `assign_h160` and `assign_u64` for `ByteDecomposition <F, N>` and then checking the correct `N` with assert at runtime is the nicest thing I can imagine.